src/OrleansRuntime/Async/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Runtime.Providers;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private readonly Interner<Guid, GrainCancellationToken> _cancellationTokens;
```

Comment:
Made interner non-static again in order to avoid the need in compound keys (grain id + token id). Can make them static again if the performance impact of having multiple timers is expected to be high enough.